android/src/main/java/io/customer/reactnative/sdk/extension/TypeConversion.kt

```diff
@@ -20,3 +25,24 @@ internal fun Double?.toCIOLogLevel(fallback: CioLogLevel = CioLogLevel.NONE): Ci
     return if (this == null) fallback
     else CioLogLevel.values().getOrNull(index = toInt() - 1) ?: fallback
 }
+
+/**
+ * Extensions function to build FCM [RemoteMessage] using RN map. This should be independent from
+ * the sender source and should be able to build a valid [RemoteMessage] for our native SDK.
+ *
+ * @param destination receiver of the message. It is mainly required for sending upstream messages,
+ * since we are using RemoteMessage only for broadcasting messages locally, we can use any non-empty
+ * string for it.
+ */
+internal fun ReadableMap.toFCMRemoteMessage(destination: String): RemoteMessage {
+    return with(RemoteMessage.Builder(destination)) {
+        getMap("data")?.toMutableStringMap()?.let { data -> setData(data) }
+        getString("messageId")?.let { id -> setMessageId(id) }
+        getString("messageType")?.let { type -> setMessageType(type) }
+        getString("collapseKey")?.let { key -> setCollapseKey(key) }
+        if (hasKey("ttl")) {
+            ttl = getInt("ttl")
+        }
+        build()
+    }
+}
```

Comment:
How did you come up with this code? I have never seen `messageId`, `messageType`, `collapseKey`, or `ttl`. Where is this data coming from and why is it important? 

Are we converting a RN FCM payload to a native FCM payload?

Reply:
Yes, that's exactly what I'm trying to do here. I tried to set all attributes required for generating [RemoteMessage using Builder object](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/RemoteMessage.Builder). The attribute names are used what [rnfirebase uses currently](https://rnfirebase.io/reference/messaging/remotemessage). As this is the most common SDK used for RN. However, if the customer has any other SDK integrated, they can still pass an object using same attribute names.

Reply:
Ok, thanks for clarifying this. 

🤔 There might be some problems with the current solution proposed in this PR. What if someone wanted to use an SDK that is not `rnfirebase` that might not have these fields present? What if the iOS or RN FCM SDKs change one of these strings, would it break out SDK? 

This PR as it currently is written feels like we are only adding official support for `rnfirebase` SDK when I think our goal is to support any FCM or APN implementation for receiving a push on a device? 

Taking a step back, our SDK's job is to take the JSON payload of a push and process it. Our SDK does not need any other fields besides the push payload. 

I want to pitch the idea of modifying this solution to be more generic where our SDK takes in the JSON payload the push service delivered to the mobile device. Then, our SDK processes it.

Reply:
The code only runs for Android so we don't have to worry about iOS payload here. The fields are retrieved with null safety operator so if there is any field missing, it should just use default value for it which will be `null` in most cases.

The fields here are decided by what `RemoteMessage.Builder` exposes. Where `RemoteMessage` is FCM class for native Android and what [our SDK already relies on heavily](https://github.com/customerio/customerio-android/blob/main/messagingpush/src/main/java/io/customer/messagingpush/CustomerIOPushNotificationHandler.kt#L44).

Lastly, I completely agree that our solution should be generic and not rely on one specific service. But the only thing we have taken from `rnfirebase` is the name of keys, which I think is okay and aligns with native `RemoteMessage.Builder`. For anyone calling this from any other SDK callbacks, they can still pass the map with same keys and this should work for them.

I hope this answers all your concerns. If there is still anything that I missed, please feel free to ask it again so I can address it.

Reply:
Good point on the keys. 

Looking through npmjs.com, I don't see a popular FCM SDK that's not `rnfirebase`. Therefore, I am much more OK with having a chance like this that's generic enough, but also using the keys that `rnfirebase` uses directly. So, the solution is coupled to `rnfirebase` loosely and that might not be a bad thing.